### DISCUSSION TITLE
What do I do if I want to cast an accessor to different data types?

### DISCUSSION THREAD

**Discussion by CaoZhongZ** [intel/llvm#3179] [Q&A]
I defined an accessor to local memory as ```int```, but then I want to use ```uint8``` accessor to manipulate the same location. Is it possible?

**Comment by al42and** [comment]
I used `sycl::buffer::reinterpret` to solve similar problem. But it requires creating a new accessor to a reinterpreted buffer. It can not reinterpret accessors.

**Comment by CaoZhongZ** [comment]
Ah, there is no ```sycl::local::reinterpret``` or something.

**Comment by alexbatashev** [comment]
@CaoZhongZ there's `cl::sycl::multi_ptr`, which is a address-space-attributed wrapper around raw pointers. It can be created from an accessor and casted to any other type. See page 140 of [SYCL Spec](https://www.khronos.org/registry/SYCL/specs/sycl-1.2.1.pdf) for more details on the interfaces. Unfortunately, there's no way to make an accessor from `multi_ptr`. SYCL 2020 would change these interfaces a bit, but there's still no direct way to cast accessors. Would this work for you at least as a workaround?
Otherwise, it makes sense to talk to @jbrodman, @mkinsner, @rolandschulz, @Pennycook, @iburyl, @rarutyun.

**Comment by CaoZhongZ** [comment]
Yes, I managed to find this part of spec and trekked over the ```reinterpret_cast``` like behavior. Finally the code would be like
```
sycl::local_ptr<uint8_t> ptr { sycl::local_ptr<void>(b_accessor) };
```
or keep the array info.

```
sycl::local_ptr<std::remove_pointer<uint8(*)[32]>::type> ptr { sycl::local_ptr<void>(b_accessor) };
```

Then compiler complaint nothing.

Of course, more expressive and succinct construct of language would be appreciated 😄
